### PR TITLE
Extract GraphiQL from core Endpoint plug

### LINKED ISSUE #16: Extract GraphiQL from core Endpoint plug
GraphiQL functionality is coupled with core graphql plug functionality should be separated

### PR DESCRIPTION
**Pull request by aweiker**
- Version that can only execute queries
- Separate version that can load GraphiQL, fixes #16 
- Allow queries to be retrieved from the server side either through a function or from a passed in option
- Upgrade to GraphiQL 0.6.1

### CODE REVIEW

File: lib/graphql/plug.ex [hunk 1]
```diff
@@ -26,18 +26,22 @@ defmodule GraphQL.Plug do
...
-  # plug GraphQL.Plug.GraphiQL
-
-  # TODO remove duplication call GraphQL.Plug.Helper.extract_init_options/1 here
   def init(opts) do
     schema = case Keyword.get(opts, :schema) do
       {mod, func} -> apply(mod, func, [])
       s -> s
     end
+
     root_value = Keyword.get(opts, :root_value, %{})
```
Comment (joshprice) [inline lib/graphql/plug.ex]: As discussed I think that the core params need to be extracted in the core `Endpoint` plug and this decorates and adds the extra behaviour -`graphiql` in this case

File: test/graphql/plug/graphiql_test.exs [hunk 2]
```diff
@@ -0,0 +1,146 @@
+defmodule GraphQL.Plug.GraphiQLTest do
```
Comment (joshprice) [inline test/graphql/plug/graphiql_test.exs:1]: This seems to duplicate the endpoint tests, can we strip this down to just graphiql specific tests?

File: lib/graphql/plug/graphiql.ex [hunk 3]
```diff
@@ -0,0 +1,109 @@
+defmodule GraphQL.Plug.GraphiQL do
+  @moduledoc """
+  This is the core plug for mounting a GraphQL server.
```
Comment (joshprice) [inline lib/graphql/plug/graphiql.ex]: docs are incorrect

File: lib/graphql/plug/graphiql.ex [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+  """
+  import Plug.Conn
+  alias Plug.Conn
+  alias GraphQL.Plug.Endpoint
+  alias GraphQL.Plug.FunctionalValue
+  alias GraphQL.Plug.Parameters
+
+  @behaviour Plug
+
+  @graphiql_version "0.4.9"
```
Comment (joshprice) [inline lib/graphql/plug/graphiql.ex]: I believe this version is out of date now, saw `0.5.0` somewhere in the last few days

File: lib/graphql/plug/graphiql.ex [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    |> put_resp_content_type("text/html")
+    |> send_resp(200, graphiql)
+  end
+
+  def use_graphiql?(%Conn{method: "GET"} = conn) do
+    accept_headers = get_req_header(conn, "accept")
+
+    case accept_headers do
+      [accept_header | _] ->
+        IO.inspect accept_header
```
Comment (joshprice) [inline lib/graphql/plug/graphiql.ex]: Inspect, think you already caught this

**Comment by joshprice** [comment]
Looks good, nice work!

As discussed, we need to name the modules here more clearly:
- [ ] `Plug`: this is the all singing, all dancing plug which does everything out of the box
- [ ] `Endpoint`: this is the core plug that should only deal with executing the query
- [ ] `Parameters`: seems like a reasonable name, maybe use singular?
- [ ] `FunctionalValue`: does root_value processing? name and role need to be clearer

File: lib/graphql/plug/graphiql.ex (hunk 4 above)
Comment (aweiker) [inline lib/graphql/plug/graphiql.ex]: Let's get that updated as a separate PR. I can pull that together in the next day or so.

Comment (joshprice) [inline lib/graphql/plug/graphiql.ex]: Sounds good

**Comment by joshprice** [comment]
Awesome work, thanks! :rocket:

**Pull request merged**